runtime/compiler/ilgen/Walker.cpp

```diff
@@ -3092,13 +3092,63 @@ TR_J9ByteCodeIlGenerator::genInvokeInterface(int32_t cpIndex)
       genInstanceof(interfaceCPIndex);
       TR::Node *instanceof = pop();
 
+      // The receiver should not be NULL at this point because there should already be a NULLCHK
+      // before the call or there is no need of NULLCHK
+      instanceof->setReferenceIsNonNull(true);
+
       TR::SymbolReference *icce =
          symRefTab()->findOrCreateIncompatibleClassChangeErrorSymbolRef(_methodSymbol);
 
       TR::Node *check =
          TR::Node::createWithSymRef(TR::ZEROCHK, 1, 1, instanceof, icce);
 
-      callTree->insertBefore(TR::TreeTop::create(comp(), check));
+      TR::TreeTop *zeroCHKTT = callTree->insertBefore(TR::TreeTop::create(comp(), check));
+
+      /*
+       * If the class is unresolved, `genInstanceof` anchors `instanceof` under a `treetop` node.
+       * Then the anchored `instanceof` treetop appears after the callTree and it commons with
+       * the previous `instanceof` node under the ZEROCHK.
+       * This causes a problem because `expandUnresolvedClassInstanceof` does not expect the
+       * `treetop` that has the anchored `instanceof` to have already been evaluated when it appears
+       * under ZEROCHK. The transformation in `expandUnresolvedClassInstanceof` will not function correctly.
+       * Therefore, the anchored `instanceof` treetop needs to be moved up before ZEROCHK.
+       *
+       * preTT
+       * ZEROCHKTT
+       *    instanceof
+       * callTree
+       * treetop
+       *    ==>instanceof
+       * nextTT
+       *
+       *    ||
+       *    ||
+       *    \/
+       *
+       * preTT
+       * treetop
+       *    instanceof
+       * ZEROCHKTT
+       *    ==>instanceof
+       * callTree
+       * nextTT
+       *
+       */
+      TR::TreeTop *instanceOfTT = callTree->getNextTreeTop();
+      if (instanceOfTT &&
+          instanceOfTT->getNode()->getOpCodeValue() == TR::treetop &&
+          instanceOfTT->getNode()->getFirstChild() &&
+          instanceOfTT->getNode()->getFirstChild() == instanceof)
+         {
+         callTree->join(instanceOfTT->getNextTreeTop());
+
+         zeroCHKTT->insertBefore(instanceOfTT);
+
+         if (comp()->getOption(TR_TraceILGen))
+            {
+            traceMsg(comp(), "%s: move the anchored instanceof n%dn before ZEROCHK n%dn\n", __FUNCTION__, instanceOfTT->getNode()->getGlobalIndex(), zeroCHKTT->getNode()->getGlobalIndex());
+            }
+         }
```

Comment:
When I implemented this type test, I decided to generate the call first and then find the call tree and insert `ZEROCHK`/`instanceof` before it. IIRC I did it that way because with OSR generating the call can also generate some pending push stores, and especially for involuntary OSR I wanted the check to be between the pending push stores and the call.

Reply:
I probably should have left a comment about that in the code...

Reply:
I experimented the possibility of generating instanceof before the callNode.

First, the generation of callNode and the generation of instanceof both push and pop nodes on the stack. The sequence of where they are generated matters [1].

In order to match what the stack should look like before the callNode is generated, I experimented by pushing the receiver node again to the stack at the beginning of genInvokeInterface and then moving the generation of the instanceof before the call node. The final stack matches the original one and the generated trees look like below:

```
   /--- trees inserted ------------------------
      n8n      (  0)  treetop                                                                              [0x7fbe1191c850] bci=[-1,6,4] rc=0 vc=0 vn=- li=- udi=- nc=1
      n7n      (  2)    instanceof  jitInstanceOf[#87  helper Method] [flags 0x400 0x0 ]                   [0x7fbe1191c800] bci=[-1,6,4] rc=2 vc=0 vn=- li=- udi=- nc=2
      n5n      (  4)      ==>aload
      n6n      (  1)      ==>loadaddr
      n13n     (  0)  NULLCHK on n5n [#32]                                                                 [0x7fbe1191c9e0] bci=[-1,6,4] rc=0 vc=0 vn=- li=- udi=- nc=1
      n12n     (  1)    PassThrough                                                                        [0x7fbe1191c990] bci=[-1,6,4] rc=1 vc=0 vn=- li=- udi=- nc=1
      n5n      (  4)      ==>aload
      n15n     (  0)  ZEROCHK [#54]                                                                        [0x7fbe1191ca80] bci=[-1,6,4] rc=0 vc=0 vn=- li=- udi=- nc=1
      n7n      (  2)    ==>instanceof
      n11n     (  0)  treetop                                                                              [0x7fbe1191c940] bci=[-1,6,4] rc=0 vc=0 vn=- li=- udi=- nc=1
      n10n     (  1)    ==>acalli
   /--- stack after ------------------------
   @1 n10n     (  1)  ==>acalli
   @0 n3n      (  1)  ==>aload
```

**However**, now we encounter the second problem that [the NULLCHK on the receiver](https://github.com/eclipse-openj9/openj9/blob/d7e7d5d374620dd6c3f30d757390ae52e9357c39/runtime/compiler/ilgen/Walker.cpp#L3088C1-L3089C1) which is separated from the call tree is after the instaceof. Then it requires either moving the NULLCHK or other changes to have the NULLCHK before the instaceof.

This fix is for a rare occurrence path with a limited impact on just this particular path. After the experiment, I'm not convinced it's worth of trying to move the generation of instanceof before the generation of the call node, which expands the change scope to the path that is not impacted by this particular problem and potentially having to move the generated IL trees as well to fix the order again. 

[1]

An example of the stack without moving where instanceof is generated: 

At the beginning of genInvokeInterface 
```
   @1 n5n      (  0)  aload  <parm 0 LSomeInterface;>
   @0 n3n      (  1)  aload  java/lang/System.out Ljava/io/PrintStream;
```
After callNode is generated 
```
   @1 n7n      (  1)  acalli  java/lang/Object.toString()Ljava/lang/String;
   @0 n3n      (  1)  aload  java/lang/System.out Ljava/io/PrintStream;
```
After pushing receiver 
```
   @2 n5n      (  3)  ==>aload
   @1 n7n      (  1)  ==>acalli
   @0 n3n      (  1)  ==>aload
```
After instanceof is generated
```
   @1 n7n      (  1)  ==>acalli
   @0 n3n      (  1)  ==>aload
```

Reply:
Thank you both for looking at it!  @a7ehuo, may I suggest expanding on your existing comments to mention the added complications that might result in trying to generate the `instanceof` earlier, and also the concern that @jdmpapin had about positioning the IL correctly with OSR enabled?  That way anyone who comes along later and tries to rearrange this code will be aware of what to watch for.

Reply:
Addressed in 2f43cd6